streaming/java/streaming-state/src/main/java/org/ray/streaming/state/strategy/CPStateStrategy.java

```diff
@@ -0,0 +1,151 @@
+package org.ray.streaming.state.strategy;
+
+import com.google.common.primitives.Longs;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.ray.streaming.state.StateException;
+import org.ray.streaming.state.StorageRecord;
+import org.ray.streaming.state.keystate.state.impl.AbstractState;
+import org.ray.streaming.state.store.IKVStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class define the checkpoint store strategy, which saves two-version data once.
+ */
+public class CPStateStrategy<V> extends AbstractStateStrategy<V> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CPStateStrategy.class);
+
+  public CPStateStrategy(IKVStore<String, Map<Long, byte[]>> backStore) {
+    super(backStore);
+  }
+
+  @Override
```

Comment:
How to use this in async state checkpoint?
Invoking sequence : 
1.  Finish
2. commit & AckCommit (another thread)
I'm not sure that's correct.
So we should make thread-safe in asyn state checkpoint.

Reply:
That is correct. In the default state strategy implementation, commit & AckCommit can be used in another thread. I'll add more comment about this.